src/target/riscv/riscv-013.c

```diff
@@ -357,10 +357,18 @@ static void select_dmi(struct target *target)
 		select_dmi_via_bscan(target);
 		return;
 	}
-	if (buf_eq(target->tap->cur_instr, select_dbus.out_value,
-				target->tap->ir_length))
-		return;
-	jtag_add_ir_scan(target->tap, &select_dbus, TAP_IDLE);
+	if (!target->tap->enabled)
+		LOG_TARGET_ERROR(target, "BUG: Target's TAP '%s' is disabled!",
+				jtag_tap_name(target->tap));
+
+	/* FIXME: make "tap" a const pointer. */
+	for (struct jtag_tap *tap = jtag_tap_next_enabled(NULL);
+			tap; tap = jtag_tap_next_enabled(tap))
+		if ((tap != target->tap && !tap->bypass)
```

Comment:
Could you please help me understand in which case there can be more than one non-bypassed TAP.

Is this pull request part of preparation for [this upstream change](https://review.openocd.org/c/openocd/+/8504/1) that adds multi-TAP scans?

If so, please consider mentioning it in the commit message.

Reply:
You are right, this is related to multi-TAP scans, in particular to [`jtag execute ...` command](https://review.openocd.org/c/openocd/+/8505).

However, I believe this to be a valid change by it's own.

To avoid adding `jtag_add_ir_scan()` it is needed to verify that the whole chain is in the expected state (since `jtag_add_ir_scan()` modifies the state of the whole chain).

Reply:
Thank you for explanation :+1: